src/libmw/src/mmr/Segment.cpp

```diff
@@ -0,0 +1,120 @@
+#include <mw/mmr/Segment.h>
+#include <mw/mmr/LeafSet.h>
+#include <mw/mmr/MMR.h>
+#include <mw/mmr/MMRUtil.h>
+
+using namespace mmr;
+
+Segment SegmentFactory::Assemble(const IMMR& mmr, const ILeafSet& leafset, const LeafIndex& first_leaf_idx, const uint16_t num_leaves)
+{
+    if (!leafset.Contains(first_leaf_idx)) {
+        return {};
+    }
+
+    Segment segment;
+    segment.leaves.reserve(num_leaves);
+
+    mmr::LeafIndex leaf_idx = first_leaf_idx;
+    mmr::LeafIndex last_leaf_idx = first_leaf_idx;
+    while (segment.leaves.size() < num_leaves && leaf_idx < leafset.GetNextLeafIdx()) {
+        if (leafset.Contains(leaf_idx)) {
+            last_leaf_idx = leaf_idx;
+            segment.leaves.push_back(mmr.GetLeaf(leaf_idx).vec());
```

Comment:
Isn't this supposed to store the leaf hashes? https://github.com/litecoin-project/litecoin/blob/970f4cf92c55a2a96f214b0f3a92529314e471ba/src/libmw/include/mw/mmr/Segment.h#L20-L21

Reply:
So the old code got confused between the hash of the leaf data, and the node hash at that leaf. We need the output ids to verify the mmr.

Reply:
Gotcha. Well, this code was written to be able to create a segment from any PMMR, not just the output one. So Leaf.vec() is not guaranteed to contain a mw::Hash, for example, Kernel MMR. Can we just change leaves to be a `std::vector<mmr:Leaf>` then, and then consumers who call it with the OutputPMMR (i.e. net_processing.cpp) can call `.vec()` when they need the output ID? Comment on `leaves` will also need updated to match.